metric/point.go

```diff
@@ -0,0 +1,237 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"time"
+
+	"go.opencensus.io/exemplar"
+)
+
+// Point is a single data point of a time series.
+type Point struct {
+	// Time is the point in time that this point represents in a time series.
+	Time time.Time
+	// ValueType is the type of value held in this point.
+	ValueType ValueType
+	// Value is the value of this point. Prefer using ReadValue to switching on
+	// the value type, since new value types might be added.
+	Value interface{}
+}
+
+//go:generate stringer -type ValueType
+
+// ValueType is the type of value held in a point.
+type ValueType int
+
+// Value types. New value types may be added.
+const (
+	ValueTypeDouble ValueType = iota
+	ValueTypeInt64
+	ValueTypeDistribution
+	ValueTypeSummary
+)
+
+// NewDoublePoint creates a new Point holding a float64 value.
+func NewDoublePoint(t time.Time, val float64) Point {
+	return Point{
+		ValueType: ValueTypeDouble,
+		Value:     val,
+		Time:      t,
+	}
+}
+
+// NewInt64Point creates a new Point holding an int64 value.
+func NewInt64Point(t time.Time, val int64) Point {
+	return Point{
+		ValueType: ValueTypeInt64,
+		Value:     val,
+		Time:      t,
+	}
+}
+
+// NewDistributionPoint creates a new Point holding a Distribution value.
+func NewDistributionPoint(t time.Time, val *Distribution) Point {
+	return Point{
+		ValueType: ValueTypeDistribution,
+		Value:     val,
+		Time:      t,
+	}
+}
+
+// NewSummaryPoint creates a new Point holding a Summary value.
+func NewSummaryPoint(t time.Time, val *Summary) Point {
+	return Point{
+		ValueType: ValueTypeSummary,
+		Value:     val,
+		Time:      t,
+	}
+}
+
+// ValueVisitor allows reading the value of a point.
+type ValueVisitor interface {
+	VisitDoubleValue(float64)
+	VisitInt64Value(int64)
+	VisitDistributionValue(*Distribution)
+	VisitSummaryValue(*Summary)
+}
+
+// ReadValue accepts a ValueVisitor and calls the appropriate method with the
+// value of this point.
+// Consumers of Point should use this in preference to switching on the type
+// of the value directly, since new value types may be added.
+func (p Point) ReadValue(vv ValueVisitor) {
```

Comment:
Is this something that exporters would use? If yes, maybe we shouldn't have it in this package.

Reply:
Yes, it is. This whole package is mostly for use by exporters.

Reply:
Is there any end to end example usage for the new APIs here? It is not clear what the package is for. If it is mainly for the exporters, we shouldn't use a top-level package like "metric"  where we would like to put the developer-facing APIs to.